src/ATokenNaive.sol

```diff
@@ -0,0 +1,357 @@
+    bytes memory params
+  ) external returns (uint256);
+}
+
+interface IVotingToken {
+  function transfer(address to, uint256 amount) external returns (bool);
+  function transferFrom(address from, address to, uint256 amount) external returns (bool);
+  function delegate(address delegatee) external;
+  function getPastVotes(address account, uint256 blockNumber) external returns (uint256);
+}
+
+contract ATokenNaive is AToken {
+  using WadRayMath for uint256;
+  using SafeCast for uint256;
+  using GPv2SafeERC20 for IERC20;
+
+  /// @notice The voting options corresponding to those used in the Governor.
+  enum VoteType {
+    Against,
+    For,
+    Abstain
+  }
+
+  /// @notice Data structure to store vote preferences expressed by depositors.
+  struct ProposalVote {
+    uint128 againstVotes;
+    uint128 forVotes;
+    uint128 abstainVotes;
+  }
+
+  /// @notice Must call castVote within this many blocks of the proposal
+  /// deadline, so-as to allow ample time for all depositors to express their
+  /// vote preferences. Corresponds to roughly 4 hours, assuming 12s/block.
+  uint32 public constant CAST_VOTE_WINDOW = 1200;
+
+  /// @notice Map depositor to deposit amount.
+  mapping(address => uint256) public deposits;
+
+  /// @notice Map borrower to total amount borrowed.
+  mapping(address => uint256) public borrowTotal;
+
+  /// @notice Map proposalId to an address to whether they have voted on this proposal.
+  mapping(uint256 => mapping(address => bool)) private _proposalVotersHasVoted;
+
+  /// @notice Map proposalId to vote totals expressed on this proposal.
+  mapping(uint256 => ProposalVote) public proposalVotes;
+
+  /// @notice The governor contract associated with this governance token.
+  IFractionalGovernor public immutable governor;
+
+  /// @dev Constructor.
+  /// @param _pool The address of the Pool contract
+  /// @param _governor The address of the flex-voting-compatable governance contract.
+  constructor(IPool _pool, address _governor) AToken(_pool) {
+    governor = IFractionalGovernor(_governor);
+  }
+
+  // TODO Is there a better way to do this? It cannot be done in the constructor
+  // because the AToken is just used a proxy -- it won't share an address with
+  // the implementation (i.e. this code).
+  function selfDelegate() public {
+    IVotingToken(governor.token()).delegate(address(this));
+  }
+
+  /// @notice Method which returns the deadline (as a block number) by which
+  /// depositors must express their voting preferences to this Pool contract. It
+  /// will always be before the Governor's corresponding proposal deadline.
+  /// @param proposalId The ID of the proposal in question.
+  function internalVotingPeriodEnd(uint256 proposalId)
+    public
+    view
+    returns (uint256 _lastVotingBlock)
+  {
+    _lastVotingBlock = governor.proposalDeadline(proposalId) - CAST_VOTE_WINDOW;
+  }
+
+  /// TODO how to handle onBehalfOf?
+  /// TODO should this revert if the vote has been cast?
```

Comment:
Would like feedback on these two items as well

Reply:
Sorry, what I meant is: should this revert if `aToken.castVote` has already been called -- e.g. if the aToken has already submitted its votes to the governor for the proposal. If it has, there would be no point for anyone to express their preference to the aToken on that proposal

Reply:
Got it. So if `aToken.castVote` is called very early, can other people express their preferences and re-vote? I'm not sure if the governor allows it or if it reverts?

But in general I think yes, revert if it's too late to keep the storage consistency, feels cleaner that way. Do we need a check on `aToken.castVote` to prevent calling until some `internalVotingPeriodStart` timestamp?

Reply:
> So if aToken.castVote is called very early, can other people express their preferences and re-vote? I'm not sure if the governor allows it or if it reverts?

No, castVote can only be called once. The governor will revert otherwise. That's why we have the explicit  period during which `castVote` _can't_ be called -- it gives people dedicated time to vote. Otherwise someone could just express their voting preference then immediately call `castVote`, effectively burning everyone else's voting power.

> I think yes, revert if it's too late to keep the storage consistency, feels cleaner that way

👍 

> Do we need a check on aToken.castVote to prevent calling until some internalVotingPeriodStart block/timestamp?

That's what [this](https://github.com/ScopeLift/flexible-voting/pull/11/files#diff-5a237783c25a06fff902b80c7117e7e15796237c85e0a05dadc7b888afdcd8ecR135-R138) does.

Reply:
Agreed—revert if the vote has already been cast to avoid confusion.

Reply:
Done in 161fd2d